0004/README.md

```diff
@@ -0,0 +1,524 @@
+-  Both are compatible with the "trust on first use" approach described
+    below, because both generate provenance expectations, based on:
+    -  A configurable set of provenance properties that are not
+        expected to change. This could be an organization-wide or even
+        ecosystem-wide default setting.
+    -  The content of a 'good' provenance instance. 
+
+I am inclined to prefer alternative 1, because it is more explicit about what
+provenance will or will not be compliant. Alternative 2 may be preferable if
+only information about the last package version needs to be remembered.
+
+## Bootstrapping policy support
+
+### Generating provenance expectations
+
+To overcome a bootstrapping problem, we assume that compromises will be rare,
+and propose to generate package-specific SLSA provenance expectations on first
+use. Each such 'trust on first use' (TOFU) policy contains expectations based on
+observed provenance, and can be used as a baseline to detect unexpected changes,
+for example a too low SLSA level, missing provenance, or an unexpected change in
+a 'sensitive' provenance property. A [proposed TOFU
+policy](#a-proposed-tofu-policy-for-a-single-npm-package) is presented below.
+
+### From local to centralized management
+
+We expect that provenance expectations will initially be managed locally where
+they are used: at the package developer before publishing to a registry, at the
+package registry, or at the package user. As experience grows and confidence
+increases, policies may be migrated to a central registry and managed there.
+
+Future work is to implement centralized policy management.
+
+### From alerting-only to enforcing
+
+We expect that initially the policy evaluator will alert for an unexpected
+change, and not block it.
+
+-  An auditing policy can be evaluated post-upload and post-installation.
+    This minimizes interference with existing npm tooling. 
+-  An enforcing policy needs to be more tightly integrated with npm tools,
+    so that it can block a non-compliant package.
+
+Future work is to integrate the policy evaluator with npm tooling.
+
+### A proposed TOFU policy for a single npm package
+
+This is for a policy that lists explicit expectations for provenance values,
+instead of expectations for what provenance fields are invariant (regardless of
+their value).
+
+Inputs: 
+
+-  Package name and version.
+-  Optional: verified package provenance.
+-  Preferences. Examples include, but are not limited to:
+    -  Names of 'sensitive' provenance properties that are not expected
+        to change such as builder ID or source repository,
+    -  Minimum SLSA level for an acceptable artifact, 
+    -  Maximum SLSA level depending on the builder. 
+
+-  Preferences may differ for different evaluator contexts: developer,
+    registry, or end user, and for a private or public registry.
+
+If no policy exists for this package, generate one with: 
```

Comment:
@wietse-gmail where you thinking we would create a separate policies file in the current directory or place this information in the lockfile somehow? Probably out of scope for this doc but curious to think about some specifics to see if I'm following here.

It also seems like we would need the ability to update this policy
- First time you see a version with provenance, create a policy entry for that package name and `>= version`
- If you later happen to install a lower version of the same package that also has provenance, update the policy version pattern to include this lower version

This would also mean the policy we can create on the registry would not necessarily be the same on you get locally when installing. Not sure this is a major problem though? Keeping a canonical policy on the registry also seems like a lot of work. Pretty sure we wouldn't want to stuff it into the [package metadata object](https://registry.npmjs.org/sigstore). 

We have been thinking of exposing a list of trusted builders from the registry that the CLI could use when verifying.

Reply:
Sorry for the late response; I was on vacation and got busy catching up.

For the initial phase I was thinking of storing auto-generated policy expectations locally, separate from the package-related files that already exist (no invasive change).

An auto-generated policy can indeed differ between developer, registry, and end user; the result reflects what package versions have been evaluated. Would there be other differences?

I agree that it makes sense to have a separate policy for what builders etc. to trust. This can be site-specific, like the preferences that I mention a little later.

Reply:
@feelepxyz Does this address your concern or should I add text that looks ahead into implementation?